website/src/utils/createDownloadAPIRoute.ts

```diff
@@ -0,0 +1,151 @@
+): APIRoute {
+    return async ({ params, redirect, request }) => {
+        const accessionVersion = params.accessionVersion!;
+
+        const isDownload = new URL(request.url).searchParams.has('download');
+
+        const result = await getSequenceData(accessionVersion, fileSuffix, undefinedVersionRedirectUrl, getData);
+        if (!result.isOk()) {
+            return new Response(undefined, {
+                status: 404,
+            });
+        }
+
+        if (result.value.type === ResultType.REDIRECT) {
+            return redirect(result.value.redirectUrl);
+        }
+
+        const headers: Record<string, string> = {
+            'Content-Type': contentType,
+        };
+        if (isDownload) {
+            const filename = `${accessionVersion}.${fileSuffix}`;
+            headers['Content-Disposition'] = `attachment; filename="${filename}"`;
+        }
+
+        return new Response(result.value.data, {
+            headers,
+        });
+    };
+}
+
+enum ResultType {
+    DATA = 'data',
+    REDIRECT = 'redirect',
+}
+
+type Data = {
+    type: ResultType.DATA;
+    data: string;
+};
+
+type Redirect = {
+    type: ResultType.REDIRECT;
+    redirectUrl: string;
+};
+
+const getSequenceDataWithOrganism = async (
+    accessionVersion: string,
+    organism: string,
+    fileSuffix: string,
+    undefinedVersionRedirectUrl: RedirectRoute,
+    getter: DataDownloader,
+): Promise<Result<Data | Redirect, ProblemDetail>> => {
+    const { accession, version } = parseAccessionVersionFromString(accessionVersion);
+
+    const lapisClient = LapisClient.createForOrganism(organism);
+
+    if (version === undefined) {
+        const latestVersionResult = await lapisClient.getLatestAccessionVersion(accession);
+        return latestVersionResult.map((latestVersion) => ({
+            type: ResultType.REDIRECT,
+            redirectUrl: undefinedVersionRedirectUrl(latestVersion),
+        }));
+    }
+
+    const dataResult: Result<string, ProblemDetail> = await getter(accessionVersion, organism);
+
+    // the result might be 'ok' but empty.
+    if (dataResult.isOk()) {
+        if (dataResult.value.trim().length === 0) {
+            return err({
+                type: 'about:blank',
+                title: 'Not Found',
+                status: 0,
+                detail: 'No data found for accession version ' + accessionVersion,
+                instance: '/seq/' + accessionVersion + `.${fileSuffix}`,
+            });
+        }
+    }
+
+    return dataResult.map((data) => ({
+        type: ResultType.DATA,
+        data,
+    }));
+};
+
+const getSequenceData = async (
+    accessionVersion: string,
+    fileSuffix: string,
+    undefinedVersionRedirectUrl: RedirectRoute,
+    getter: DataDownloader,
+) => {
+    // We don't know which organism the accessionVersion belongs to,
+    // so we just try all of them until we get a success.
+    const organisms = getConfiguredOrganisms();
+    const results = await Promise.all(
```

Comment:
Not blocking because it's an existing issue but while reading I noticed that this `Promise.all` isn't ideal. If one LAPIS instance for one organism is unhealthy and takes 10 secs to respond it will make all organisms take 10 seconds to respond. We use better logic here: https://github.com/loculus-project/loculus/blob/a04412f5feca4607961c512968f65ab0be4ace09/website/src/pages/seq/%5BaccessionVersion%5D/findOrganismAndData.ts#L6 . If you are willing to do that while touching this code, great, if not we can make an issue

Reply:
Thanks for the pointer, I've made the change: [be67e74](https://github.com/loculus-project/loculus/pull/3003/commits/be67e742bbdb9079a26e8c9fa6b87a0d1e2de066)